### PR TITLE
Add additional unit tests

### PR DESCRIPTION
**Pull request by liz-is**
This PR replaces #31 - the important commits have been cherry-picked and applied on top of `dev`.

@ChimaStan, @enricoparisini and @cpiazz have added tests for:

* utils
* filters
* io
* feature extractor

### CODE REVIEW

**Comment by liz-is** [comment]
The tests don't currently run during the Github Actions workflow as there's an error collecting the tests to run them!

```
==================================== ERRORS ====================================
____________________ ERROR collecting test/test_patching.py ____________________
ImportError while importing test module '/home/runner/work/pyslyde/pyslyde/test/test_patching.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.13.11/x64/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
test/test_patching.py:9: in <module>
    from pyslide.patching import Patching, Slide
E   ModuleNotFoundError: No module named 'pyslide'
```

The issue is with the line `from pyslide.patching import Patching, Slide` in `test/test_patching.py`. Updating `pyslide`-->`pyslyde` won't be sufficient to fix this, as there is no longer a `patching` module. 

* `Slide` seems to be relocated to `pyslyde.slide`
* `Patching` doesn't appear to exist any more - has this been renamed?

**Q:** if `Patching` doesn't exist any more, can `test/test_patching.py` be deleted?

**Comment by liz-is** [comment]
I have run the tests locally using `python -m pytest --ignore test/test_patching.py`. This should work for everyone - let me know if not.

The results of this on my machine are: `16 failed, 55 passed, 6 skipped, 1 warning in 11.32s`

The failing tests on my machine can be divided into different categories:

**Missing test input files:**

```
FAILED test/test_pyslide.py::TestSlide::test_detect_components - FileNotFoundError: [Errno 2] No such file or directory: '14.90610 C L2.11.json'
FAILED test/test_pyslide.py::TestSlide::test_generate_annotations - FileNotFoundError: [Errno 2] No such file or directory: '14.90610 C L2.11.json'
FAILED test/test_pyslide.py::TestSlide::test_generate_region - FileNotFoundError: [Errno 2] No such file or directory: '14.90610 C L2.11.json'
FAILED test/test_pyslide.py::TestSlide::test_get_border - FileNotFoundError: [Errno 2] No such file or directory: '14.90610 C L2.11.json'
FAILED test/test_pyslide.py::TestSlide::test_resize_border - FileNotFoundError: [Errno 2] No such file or directory: '14.90610 C L2.11.json'
FAILED test/test_pyslide.py::TestSlide::test_slide_mask - FileNotFoundError: [Errno 2] No such file or directory: '14.90610 C L2.11.json'
FAILED test/test_pyslide.py::TestAnnotations::test_asap - FileNotFoundError: [Errno 2] No such file or directory: 'data/annotations/asap_annotations.xml'
FAILED test/test_pyslide.py::TestAnnotations::test_csv - FileNotFoundError: [Errno 2] No such file or directory: 'data/annotations/csv_annotations.csv'
FAILED test/test_pyslide.py::TestAnnotations::test_imagej - FileNotFoundError: [Errno 2] No such file or directory: 'data/annotations/imagej_annotations.xml'
FAILED test/test_pyslide.py::TestAnnotations::test_json - FileNotFoundError: [Errno 2] No such file or directory: 'data/annotations/json_annotations.json'
```

@gregoryverghese I think you wrote the tests in `test/test_pyslide.py` - could you share these missing input files? I think ideally we should add these in a folder within `test/` called something like `test/test_data/`. However, if these are _large_ files, we should avoid adding them to the repo and maybe read them from another online location instead. 

**Problems with input data?:**

```
FAILED test/test_filters.py::test_image_entropy - cv2.error: OpenCV(4.12.0) /Users/xperience/GHA-Actions-OpenCV/_work/opencv-python/opencv-python/opencv/modules/imgproc/src/color.simd_helpers.hpp:92: error: (-15:Bad number of channels) in f...
FAILED test/test_filters.py::test_entropy - cv2.error: OpenCV(4.12.0) /Users/xperience/GHA-Actions-OpenCV/_work/opencv-python/opencv-python/opencv/modules/imgproc/src/color.simd_helpers.hpp:92: error: (-15:Bad number of channels) in f...
```

Full error:

```
E       cv2.error: OpenCV(4.12.0) /Users/xperience/GHA-Actions-OpenCV/_work/opencv-python/opencv-python/opencv/modules/imgproc/src/color.simd_helpers.hpp:92: error: (-15:Bad number of channels) in function 'cv::impl::(anonymous namespace)::CvtHelper<cv::impl::(anonymous namespace)::Set<3, 4>, cv::impl::(anonymous namespace)::Set<1>, cv::impl::(anonymous namespace)::Set<0, 2, 5>>::CvtHelper(InputArray, OutputArray, int) [VScn = cv::impl::(anonymous namespace)::Set<3, 4>, VDcn = cv::impl::(anonymous namespace)::Set<1>, VDepth = cv::impl::(anonymous namespace)::Set<0, 2, 5>, sizePolicy = cv::impl::(anonymous namespace)::NONE]'
E       > Invalid number of channels in input image:
E       >     'VScn::contains(scn)'
E       > where
E       >     'scn' is 1
```

It's not clear to me what's going on here, anyone have any ideas?

**Problems with arguments:**

```
FAILED test/test_utils.py::test_oneHotToMask_output - TypeError: oneHotToMask() got an unexpected keyword argument 'background'
FAILED test/test_utils.py::test_entropy_functions - NameError: name 'disk' is not defined
FAILED test/test_utils.py::test_detect_tissue - cv2.error: OpenCV(4.12.0) :-1: error: (-5:Bad argument) in function 'resize'
FAILED test/test_utils.py::test_border - AttributeError: 'numpy.ndarray' object has no attribute 'dimensions'
```

This looks to me a bit like the versions of dependencies that I have installed (based on what's specified in the package requirements) might be different from the versions that were used to develop these tests. @cpiazz do these tests pass for you? maybe we should compare environments and see if there's discrepancies? 

**Skipped tests:**

6 tests in `test/test_io.py` are skipped because `rocksdb (and rocksdb_io import) not available`. 

I was unable to install `rocksdb` locally, due to a compiler error. The [github repo](https://github.com/twmht/python-rocksdb) has been archived, so it seems like this is now an abandoned package :( There appear to be [other `rocksdb`-related packages in PyPI](https://pypi.org/search/?q=rocksdb), but I'm not sure if any of these are a suitable replacement. Is this essential functionality for PySlyde, or should we remove it if there's no supported package? @enricoparisini you're the test author @gregoryverghese you're the author of the original `rocksdb_io.py` code - what do you think?

NB I don't think this is an issue that needs to be solved in this PR, but I did want to flag up why these tests are being skipped!

**Conclusion:**

I think Chima has said he's happy to lead on getting all the tests passing, and I'm happy to help with the technical side of things. But we'll definitely need your input to get this done, so thanks in advance for your help! 😊

File: README.md [hunk 1]
```diff
@@ -21,6 +21,14 @@ PySlyde is a comprehensive Python package for preprocessing pathology whole slid
 
 ## Installation
 
+## System requirements
```
Comment (liz-is) [inline README.md:24]: Non-blocking: NB that in general it's good practice to do this kind of docs updates as a separate branch. Having a separate branch means that the docs updates could be merged sooner without having to wait for the tests updates to be complete!

This suggestion will work on some Linux systems if the user has admin rights, but not if they lack admin rights or are working on Windows or Mac. We should note this, and ideally provide instructions for other OSes. Probably it's easiest to refer to the official [OpenSlide docs](https://openslide.org/).

File: pyslyde/encoders/feature_extractor.py [hunk 2]
```diff
@@ -23,12 +23,10 @@
 from torchvision import transforms as T
 
 from pyslyde.encoders.ctran import ctranspath
-#from pyslyde.encoders.HistoSSLscaling.rl_benchmarks.models import iBOTViT 
-#from HIPT_4K.hipt_model_utils import get_vit256, get_vit4k
-#from pyslyde.encoders.HIPT.HIPT_4K.hipt_model_utils import eval_transforms
-#from pyslyde.encoders.HIPT.HIPT_4K import vision_transformer as vits
-#from pyslyde.encoders.HIPT.HIPT_4K.hipt_4k import HIPT_4K
-#from lmdb_data import LMDBRead, LMDBWrite
+# from pyslyde.encoders.HistoSSLscaling.rl_benchmarks.models import iBOTViT
```
Comment (liz-is) [inline pyslyde/encoders/feature_extractor.py]: Non-blocking: Ideally unneeded code should be removed rather than commented out

File: pyslyde/encoders/feature_extractor.py [hunk 3]
```diff
@@ -113,82 +111,95 @@ def _ciga(self):
...
 
-    def _hipt4k(self):
-        model = HIPT_4K()
-        model.eval()
-        self.transforms = eval_transforms()
-        return model
-
-
-    def _hipt256(self):
+    # def _hipt4k(self):
```
Comment (liz-is) [inline pyslyde/encoders/feature_extractor.py]: Non-blocking: as above, ideally this would be removed rather than commented out if no longer needed

File: pyslyde/encoders/feature_extractor.py [hunk 4]
```diff
@@ -237,15 +231,12 @@ def _uni(self):
...
         """
-        see https://huggingface.co/paige-ai/Virchow2 
+        See https://huggingface.co/paige-ai/Virchow2 
         """
-
-        #login(os.getenv('HUGGINGFACE_TOKEN'))  # HUGGINGFACE_TOKEN is an environment variable
-        login(os.getenv('HUGGINGFACE_TOKEN'), add_to_git_credential=True)  # To save token to your Git credentials
-
-        # need to specify MLP layer and activation function for proper init
+        login(os.getenv('HUGGINGFACE_TOKEN'))  # HUGGINGFACE_TOKEN is an environment variable
```
Comment (liz-is) [inline pyslyde/encoders/feature_extractor.py]: Non-blocking, can revisit later: Should this be done outside of the function, rather than logging in to Huggingface as a side effect of calling this function? it would allow the user more control over how login is done

File: test/test_feature_extractor.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    assert encoder_name in str(fg.encoder), f"Expected encoder '{encoder_name}' but got {type(fg.encoder)}"
+
+
+def test_invalid_encoder_name_raises_keyerror(mock_checkpoint):
+    with pytest.raises(KeyError):
+        FeatureGenerator(model_name="moco", model_path=str(mock_checkpoint), encoder_name="invalid_encoder").encoder
+
+
+# Model Loader Tests
+@pytest.mark.parametrize(
```
Comment (liz-is) [inline test/test_feature_extractor.py]: Nice use of parametrization!

File: test/test_filters.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+import cv2
+from skimage.filters.rank import entropy as sk_entropy
+from skimage.morphology import disk
+
+from pyslyde.util import filters
+import os
+from contextlib import redirect_stdout
+
+
+def make_rgb_image(color=(128, 128, 128), size=(64, 64)):
```
Comment (liz-is) [inline test/test_filters.py]: Non-blocking: this could be a [pytest fixture](https://docs.pytest.org/en/stable/explanation/fixtures.html#about-fixtures), as could `DummyPatch`, I think.

File: test/test_filters.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    result = filters.image_entropy(img)
+    
+    # Check types and expected range
+    assert isinstance(result, float)
+    assert 0 <= result < 8  # typical entropy range for 8-bit images
+    
+    # For uniform image, mean entropy should be near zero
+    assert result < 1e-3, f"Expected near-zero entropy, got {result}"
+    
+def test_entropy():
```
Comment (liz-is) [inline test/test_filters.py]: I think it would be clearer to split this function up into separate tests for the general and above/below threshold cases.

There are a few other tests in this file that I think could also be clearer if split up into smaller functions.

File: test/test_filters.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+    # channel threshold
+    tile = make_rgb_image(color=(10, 200, 10))
+    # channel 1 (G) mean is 200
+    assert filters.tile_intensity(tile, threshold=100, channel=1) is True
+    assert filters.tile_intensity(tile, threshold=250, channel=1) is None
+    
+    assert filters.tile_intensity(tile, threshold=50, channel=0) is None
+    
+def test_remove_black():
```
Comment (liz-is) [inline test/test_filters.py]: Similarly, I think this would be clearer if split into separate tests for the different cases.

File: test/test_filters.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+    # keeps non blue patches
+    yellow = make_rgb_image(color=(255, 255, 0))
+    patch = DummyPatch([yellow])
+    result = filters.remove_blue(patch)
+    assert len(result._patches) == 1
+    assert result._patches[0] is yellow # patch is exactly the original yellow object
+    
+    
+# ---------- UNITTST RUNNER ----------
```
Comment (liz-is) [inline test/test_filters.py]: The block from here is unneeded if we run the tests with `pytest`

File: test/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,676 @@
...
+    np.testing.assert_allclose(rgb, expected_rgb, atol=1e-6)
+    
+
+    # Check shape
+    assert rgb.shape == (3, 3, 3)
+
+    # Should not be all zeros
+    assert not np.all(rgb == 0)
+    
+    # Edge case: mask contains only background
```
Comment (liz-is) [inline test/test_utils.py:118]: It could be clearer if these edge cases are split into separate tests - there are a few other functions in this file that I think could also be clearer if split up into separate tests.

File: test/test_utils.py [hunk 11]
```diff
@@ -0,0 +1,676 @@
...
+         patch('numpy.load', side_effect=fake_arrays), \
+         patch('builtins.print'):
+
+        # Call function
+        ipca = utilities.get_pca()
+
+        # Should return IncrementalPCA
+        assert isinstance(ipca, IncrementalPCA)
+
+def test_sample_patches():
```
Comment (liz-is) [inline test/test_utils.py]: I think pytest fixtures would be helpful here for the dummy classes!

This test function would also benefit from being split into smaller functions, I think.

File: test/test_utils.py [hunk 12]
```diff
@@ -0,0 +1,676 @@
...
+    # Call the property
+    thumbnail = td.tissue_thumbnail
+    
+    # Assertions
+    assert isinstance(thumbnail, np.ndarray), "Thumbnail should be a numpy array"
+    assert thumbnail.shape[2] == 3, "Thumbnail should have 3 channels (RGB)"
+    assert np.all(thumbnail >= 0) and np.all(thumbnail <= 255), "Pixel values should be in 0-255 range"
+    
+    
+# ---------- UNITTST RUNNER ----------
```
Comment (liz-is) [inline test/test_utils.py]: As above, this block isn't needed if we run tests with `pytest`

File: test/test_io.py [hunk 13]
```diff
@@ -0,0 +1,753 @@
...
+except Exception:
+    RocksNpyObject = RocksDBWrite = RocksDBRead = None
+    HAS_ROCKSDB = False
+
+
+##################################################################
+# Helpers
+##################################################################
+
+def _run_and_report(label: str, suite: unittest.TestSuite) -> unittest.result.TestResult:
```
Comment (liz-is) [inline test/test_io.py]: This isn't needed if running tests with pytest. However if running only one subset of tests would be a common task it might be easier to split them into separate files?

File: test/test_io.py [hunk 14]
```diff
@@ -0,0 +1,753 @@
...
+    if result.wasSuccessful():
+        print(f"\n[OK] Finished tests for {label}")
+    else:
+        f = len(result.failures)
+        e = len(result.errors)
+        s = len(getattr(result, "skipped", []))
+        print(f"\n[FAIL] Finished tests for {label} (failures={f}, errors={e}, skipped={s})")
+    return result
+
+def _synthetic_parser(coords, shape=(8, 8, 3), dtype=np.uint8):
```
Comment (liz-is) [inline test/test_io.py:57]: Could possibly be a pytest fixture, although I don't know how well that would integrate with the use of `unittest` features in this set of tests!

File: test/test_io.py [hunk 15]
```diff
@@ -0,0 +1,753 @@
...
+        self.assertGreaterEqual(len(self._collect_tfrecords('validation')), 1)
+        self.assertGreaterEqual(len(self._collect_tfrecords('test')), 1)
+
+
+##################################################################
+# Main: run Disk, LMDB, RocksDB, TFRecords (class), then tfrecord_write (functions)
+#       with per-section summaries.
+##################################################################
+
+if __name__ == "__main__":
```
Comment (liz-is) [inline test/test_io.py]: This block is not needed if running tests with pytest

File: test/test_io.py [hunk 16]
```diff
@@ -0,0 +1,753 @@
...
+            val = (x + 2.5 * y) / 255.0
+            tile = np.full(shape, val, dtype=dtype)
+        yield (x, y), tile
+
+
+##################################################################
+# Tests for pyslyde/io/disk_io.py
+##################################################################
+
+class TestDiskWrite(unittest.TestCase):
```
Comment (liz-is) [inline test/test_io.py:77]: I'm not super familiar with `unittest` test writing style, but I found these all pretty clear and well organised into reasonably-sized tests - nice work!

File: pyslyde/encoders/feature_extractor.py
Comment (liz-is) [inline pyslyde/encoders/feature_extractor.py:1]: Generally it's good practice to separate functional changes from changes to just the tests. In this case it might have been a good idea to create a branch for work on `feature_extractor.py` that included both functional changes and the tests for this module, but separate from the utils/io/filters tests. Not a blocking problem here though, just a suggestion for the future!

**Review by liz-is — COMMENTED** [review]
This is an initial review mainly focusing on the test organisation, and not reviewing any of the functionality as I'm not sufficiently familiar with the research domain.

Generally the tests appear pretty thorough - I'll try to add in a coverage calculation to help highlight any areas where more tests could be added. There's some nice usage of features like fixtures and parametrization! However, in some places the test functions are quite long, and the test results could be clearer if these were split up into separate tests for the different test cases (maybe using parametrization).

None of these comments are blocking, just things to think about!

File: test/test_filters.py (hunk 6 above)
Comment (cpiazz) [reply test/test_filters.py]: Added @pytest.fixture

File: test/test_filters.py (hunk 7 above)
Comment (cpiazz) [reply test/test_filters.py]: Functions split into separate tests

File: test/test_filters.py (hunk 9 above)
Comment (cpiazz) [reply test/test_filters.py]: Removed

File: test/test_utils.py (hunk 10 above)
Comment (cpiazz) [reply test/test_utils.py:118]: Functions split into separate tests

File: test/test_utils.py (hunk 12 above)
Comment (cpiazz) [reply test/test_utils.py]: Removed

**Comment by liz-is** [comment]
Thanks @enricoparisini and @cpiazz for your work on this! It's looking good :) just a few more to go!

@ChimaStan are you able to coordinate with @gregoryverghese to get the missing test data and decide what to do with `test/test_patching.py`?

**Comment by gregoryverghese** [comment]
We can remove test_patching.py this was a old module.

For `test_pyslide` I have uploaded datasets [here](https://drive.google.com/drive/folders/1fkr3kGlfwpZiv8i3Xfzl29OaUtvR2fj-?usp=drive_link). Although @ChimaStan I think we might have to use a new json and ndpi file for  `setUpClass()` as the json format for annotations is very old and might not parse correctly with the latest method.

File: .github/workflows/python-package.yml [hunk 17]
```diff
@@ -37,9 +37,9 @@ jobs:
         python -m pip install --upgrade pip openslide-bin
         df -h
         python -m pip install .[dev]
-    - name: Test with pytest
+    - name: Test with pytest (exclude integration-marked tests)
       run: |
-        python -m pytest
+        python -m pytest -m "not integration" --ignore=test/test_pyslide.py --ignore=test/test_patching.py
```
Comment (liz-is) [inline .github/workflows/python-package.yml:42]: Could we please delete these old/extraneous tests rather than just ignoring them? It will be tidier, and leaving these temporary `--ignore` statements in risks them not being noticed later if/when these tests are added back in, so that the tests get accidentally skipped. If someone needs to refer to the old test code later, they can check the git history.

File: pyslyde/util/utilities.py [hunk 18]
```diff
@@ -166,7 +177,14 @@ def __init__(self, slide: Union[str, OpenSlide]) -> None:
         Args:
             slide: Path to slide or OpenSlide object.
         """
-        self.slide = OpenSlide(slide) if isinstance(slide, str) else slide
+        
+        if isinstance(slide, str):
+            self.slide = OpenSlide(slide)
+        elif (isinstance(slide, OpenSlide) or isinstance(slide, np.array)):
```
Comment (liz-is) [inline pyslyde/util/utilities.py]:
```suggestion
        elif (isinstance(slide, OpenSlide) or isinstance(slide, np.ndarray)):
```

I believe the tests that are failing with `TypeError: isinstance() arg 2 must be a type, a tuple of types, or a union` are correctly picking up an issue here! `np.array` is a function to create an array, I believe the correct type to use here is `np.ndarray`.

File: test/test_utils.py [hunk 19]
```diff
@@ -5,21 +5,76 @@
...
+            self.mag_level = mag_level
+            self.border = border
+            self.step = step
+            self._patches = [f"patch_{i}" for i in range(10)]
+            self.patches = []
+    return DummyPatch("slide_1", (64, 64), 20, 4, 2)
+
+
+@pytest.fixture
+def dummy_slide(as_pil=True):
```
Comment (liz-is) [inline test/test_utils.py:46]: I made the change suggested above to `util/utilities.py` and tested locally, but got `TypeError: Slide must be of type OpenSlide, numpy array or string path to OpenSlide` for some tests. I believe this is because `dummy_slide()` returns an object of class `'unittest.mock.MagicMock'`, not a numpy array. I'm not familiar with MagicMock so I'm not sure what the fix is here, sorry!

**Comment by liz-is** [comment]
@ChimaStan could you add instructions on how to run the integration tests to `CONTRIBUTING.md`, alongside the other instructions on how to run tests? Thanks!

I ran the tests locally (without doing any additional setup like setting a Huggingface token), and while most integration tests were correctly skipped,`test_gated_model_cache_hit_does_not_require_token` ran and failed (with `huggingface_hub.errors.GatedRepoError: 401 Client Error.`) - should this be skipped too?

File: pyslyde/util/utilities.py (hunk 18 above)
Comment (cpiazz) [reply pyslyde/util/utilities.py]: I did try replacing np.array with np.ndarray in the isinstance check, but I was still running into errors in the tests because dummy_slide is a MagicMock.
I ended up relaxing the check to accept any object that exposes a shape attribute. This should covers real NumPy arrays while allowing the mocked slide object to pass through as intended.

File: test/test_utils.py (hunk 19 above)
Comment (cpiazz) [reply test/test_utils.py:46]: Solved by relaxing the check to accept any object that exposes a shape attribute. This should covers real NumPy arrays while allowing the mocked slide object to pass through as intended.

File: .github/workflows/python-package.yml (hunk 17 above)
Comment (ChimaStan) [reply .github/workflows/python-package.yml:42]: I have already removed the tests elsewhere on the branches that attend to the target modules. I've also removed ``--ignore`` from CI on the ``refactor-slide`` branch, where the tests have been replaced.

**Comment by ChimaStan** [comment]
> @ChimaStan could you add instructions on how to run the integration tests to `CONTRIBUTING.md`, alongside the other instructions on how to run tests? Thanks!
> 
> I ran the tests locally (without doing any additional setup like setting a Huggingface token), and while most integration tests were correctly skipped,`test_gated_model_cache_hit_does_not_require_token` ran and failed (with `huggingface_hub.errors.GatedRepoError: 401 Client Error.`) - should this be skipped too?

@liz-is That test is designed to verify that login is not required when a model is already cached locally.  It should not attempt to access the Hugging Face Hub.

The failure you observed indicated that the real model-loading path was still being triggered during the test. 
This has now been addressed by stubbing the model loader methods, so the test no longer attempts to download or authenticate.

Also, large FMs could previously trigger OS-level OOM kills during integration testing. 
The tests have now been updated to dynamically check available CPU RAM (and GPU VRAM if applicable) and skip these tests when insufficient memory is available.

Could you check whether error still occurs with the the updated tests on the ``refactor-feature-extractor`` branch?

I'll update CONTRIBUTING.md shortly.

File: .github/workflows/python-package.yml (hunk 17 above)
Comment (liz-is) [reply .github/workflows/python-package.yml:42]: Okay, fab, as long as this is being addressed somewhere that's great!

**Comment by liz-is** [comment]
@ChimaStan did you close this by accident?

I think we are ready to merge, just let me know if you wanted to add any additional integration tests instructions first?

**Comment by liz-is** [comment]
I can't approve this as I opened the PR, so will need someone else to do it!

**Comment by ChimaStan** [comment]
> @ChimaStan did you close this by accident?
> 
> I think we are ready to merge, just let me know if you wanted to add any additional integration tests instructions first?

@liz-is , I closed it by accident. Please see my reply.

**Comment by ChimaStan** [comment]
@liz-is , could you confirm that the ``huggingface_hub.errors.GatedRepoError: 401 Client Error`` error you mentioned has been resolved?

**Comment by liz-is** [comment]
> @liz-is , could you confirm that the `huggingface_hub.errors.GatedRepoError: 401 Client Error` error you mentioned has been resolved?

It's still failing in this branch, but I don't get the error in the `refactor-feature-extractor` branch where you've made that change. I'm not sure why you opened a new PR for that rather than just making the change on this branch, but as long as that also gets merged in sooner rather than later it's fine.

**Comment by ChimaStan** [comment]
> > @liz-is , could you confirm that the `huggingface_hub.errors.GatedRepoError: 401 Client Error` error you mentioned has been resolved?
> 
> It's still failing in this branch, but I don't get the error in the `refactor-feature-extractor` branch where you've made that change. I'm not sure why you opened a new PR for that rather than just making the change on this branch, but as long as that also gets merged in sooner rather than later it's fine.

The tests were failing on this branch, even after excluding the known problematic legacy modules. So I moved the tests to the respective branches of the target modules.

**Comment by liz-is** [comment]
I think it's better to fix the tests in the branch where they are failing, rather than make a new branch or reactivate a branch that has already been merged. We will need to ensure that the new change in #44 is compatible with the changes in this branch anyway, as they should both be merged soon, and it would be easier to check that if the changes were made on the same branch. But if you prefer this way, we can do it in two PRs. It just adds an extra step to check that PR after this one is merged, and then approve it.

**Comment by ChimaStan** [comment]
> I think it's better to fix the tests in the branch where they are failing, rather than make a new branch or reactivate a branch that has already been merged. We will need to ensure that the new change in #44 is compatible with the changes in this branch anyway, as they should both be merged soon, and it would be easier to check that if the changes were made on the same branch. But if you prefer this way, we can do it in two PRs. It just adds an extra step to check that PR after this one is merged, and then approve it.

I'm happy for both to be merged together.

**Review by ChimaStan — APPROVED** [review]
